ddclient.in

```diff
@@ -7940,6 +7957,447 @@ sub nic_infomaniak_update {
+        $request_header_string = $request_header_string.(lc $header_key).":".($compiled_headers{$header_key});
+
+        if ($header_key ne $compiled_headers_key[$#compiled_headers_key]) {
+            $request_header_string = $request_header_string."\n";
+        }
+    }
+
+    my %curl_opts = (
+        url => get_encoded_url($url),
+        method => $request_method,
+        headers => $request_header_string
+    );
+
+    if ($request_method =~ "PUT|POST|PATCH") {
+        debug("Body: ".$request_payload."\n");
+        $curl_opts{data} = $request_payload;
+    }
+
+    return geturl(%curl_opts);
+}
+
+my $AWS_ROUTE53_BASE = "https://route53.amazonaws.com/2013-04-01/";
+sub get_route53_api_url {
+    my (
+        $resource,
+        $query_parameters
+    ) = @_;
+
+    my $api_request_url = $AWS_ROUTE53_BASE.$resource;
+
+    if (defined($query_parameters)) {
+        $api_request_url .= "?".$query_parameters
+    }
+
+    return $api_request_url;
+}
+
+sub update_route53_one {
+    my (
+        $hosted_zone_id,
+        $h,
+        $resource_set_type,
+        $ip,
+        $ttl,
+        $aws_access_key_id,
+        $aws_secret_access_key,
+        $aws_region
+    ) = @_;
+
+    my $resource_type = $resource_set_type // ($ip =~ /\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}/ ? "A" : "AAAA");
+    my $ttl_to_use = $ttl // 300;
+
+    my $ROUTE53_NS = "https://route53.amazonaws.com/doc/2013-04-01/";
+    my $request_xml =
+    "<?xml version=\"1.0\" encoding=\"UTF-8\"?>
+    <ChangeResourceRecordSetsRequest xmlns=\"$ROUTE53_NS\">
+        <ChangeBatch>
+            <Changes>
+                <Change>
+                    <Action>UPSERT</Action>
+                    <ResourceRecordSet>
+                        <Name>$h</Name>
+                        <Type>$resource_set_type</Type>
+                        <TTL>$ttl_to_use</TTL>
+                        <ResourceRecords>
+                            <ResourceRecord>
+                                <Value>$ip</Value>
+                            </ResourceRecord>
+                        </ResourceRecords>
+                    </ResourceRecordSet>
+                </Change>
+            </Changes>
+        </ChangeBatch>
+    </ChangeResourceRecordSetsRequest>";
+
+    my $region = $aws_region // $ENV{"AWS_REGION"} // "us-east-1";
+    my $reply = create_signed_request((
+        service => "route53",
+        region => $region,
+        request_url => get_route53_api_url("hostedzone/".$hosted_zone_id."/rrset/"),
+        request_method => "POST",
+        request_headers => {
+            "content-type" => "application/xml",
+        },
+        request_payload => $request_xml,
+        aws_secret_access_key => $aws_secret_access_key,
+        aws_access_key_id => $aws_access_key_id
+    ));
+
+    # No response, declare as failed
+    if (!defined($reply) || !$reply) {
+        failed("KEYSYSTEMS updating %s: Could not connect to %s.", $h, $config{$h}{'server'});
```

Comment:
The error message references "KEYSYSTEMS". Is this a copy-paste error from another part of the code? It might be more appropriate to tailor this message specifically for Route53.